Add Complete Tutorials (#53)
fortran-lang/http-client: tutorial/real_proj/github-org-analyzer.md

# **Building a [GitHub Organization Analyzer](https://github.com/rajkumardongre/github-org-analyzer) in Fortran, using `http-client`** 🚀

In this tutorial, we'll create a simple Fortran program that uses the [GitHub API](https://docs.github.com/en/rest?apiVersion=2022-11-28) to retrieve and display all the repositories of the [`fortran-lang`](https://github.com/fortran-lang) organization. We'll use the [`http-client`](https://github.com/fortran-lang/http-client) and [`json-fortran`](https://github.com/jacobwilliams/json-fortran) libraries to make API requests and handle JSON responses.

# **Prerequisite** 🚩

> Before building the GitHub Organization Analyzer library and running the program, you need to ensure that you have [`fpm`](https://fpm.fortran-lang.org/) (Fortran Package Manager) installed. Additionally, there is one dependency required for the [`http-client`](https://github.com/fortran-lang/http-client) library used in the project. Follow the steps below to set up your environment:

Reviewer: Mention the minimum version of `fpm` also here.

### **Step 1: Install fpm**

[`fpm`](https://fpm.fortran-lang.org/) is the Fortran Package Manager used for building and managing Fortran projects. If you do not currently have `fpm` installed, you can follow the installation instructions available on the official `fpm` repository to install version v0.9.0 or a more recent version. The installation guide can be found here: [Installation Guide](https://fpm.fortran-lang.org/install/index.html).

### **Step 2: Install libcurl Development Headers**

> The `http-client` library, requires the libcurl development headers to be installed. On Ubuntu-based systems, you can install the required dependencies using the following command:

Reviewer: ```suggestion
The `http-client` library requires the libcurl development headers to be installed. On Ubuntu-based systems, you can install the required dependencies using the following command:
```

> The `http-client` library, requires the libcurl development headers to be installed. On Ubuntu-based systems, you can install the required dependencies using the following command:

Reviewer: Maybe:
```suggestion
The `http-client` library, requires the ``libcurl`` development headers to be installed. On Ubuntu-based systems, you can install the required dependencies using the following command:
```
?

```
sudo apt install -y libcurl4-openssl-dev
```

This command will install the necessary development headers for libcurl, enabling the `http-client` library to make API requests to fetch data from the GitHub API.

Once you have `fpm` installed and the required dependencies set up, you are ready to proceed with building and running the GitHub Organization Analyzer project.🙌

# **Let's Start Building** 👨‍💻

### **Step 1: Set up the Project**

>**Note : This requires at least fpm v0.9.0.**

1. Open your terminal or command prompt and create a new directory for the project:

> ```

Reviewer: ```suggestion
```bash
```
??

fpm new github-org-analyzer
cd github-org-analyzer
```

2. The project structure will look like this:

```
.
├── README.md
├── app
│   └── main.f90
├── fpm.toml
├── src
│   └── github-org-analyzer.f90
└── test
    └── check.f90
```

### **Step 2: Add Dependencies to `fpm.toml`**

Open the `fpm.toml` file and add the following dependencies:

```toml
[dependencies]
http = { git = "https://github.com/fortran-lang/http-client.git" }
stdlib = "*"
json-fortran = { git = "https://github.com/jacobwilliams/json-fortran.git" }
```

### **Step 3: Build the Project**

Run the following command to build the project:

```
fpm build

> ```

Reviewer: Is this step really needed here?

Author: In my understanding, when executing the `fpm run` command, the project will be built automatically during that process. Therefore, it seems unnecessary to include a separate step for building the project. So Should I remove this step?

> ### **Step 5: Import the Libraries**

Reviewer: ```suggestion
### **Step 4: Import the Libraries**
```

Open the `github-org-analyzer.f90` file in the `src` folder and import the required libraries:

```fortran
module github_org_analyzer
    use json_module, only : json_file
    use http, only : request, response_type
    use stdlib_strings, only : to_string

    ! ... (subroutine to be added later)
end module github_org_analyzer
```
* `json_module` : This module provides functionalities for parsing JSON content, and we use it here to work with JSON data obtained from the GitHub API.

* The `http` module enables us to make API calls and send HTTP requests to fetch data from the GitHub API.

* `to_string` : We use this function to convert integers to their string representations

> ### **Step 6: Create the `print_org_repositories` Subroutine**

Reviewer: ```suggestion
### **Step 5: Create the `print_org_repositories` Subroutine**
```

      !!  flag for weather the current repo found or not

      ! Make an HTTP GET request to the API URL and store the response in the `response` variable
      response = request(url=api_url)

      ! Checking for any errors. If an error occurs during the HTTP call, the `ok`
      ! attribute will be set to .false., and the `err_msg` attribute will contain
      ! the reason for the error.
      if(.not. response%ok) then
         print *, 'Request Fail: ', response%err_msg
      else
         print *, 'Fortran lang All repositories:'

         ! Initialize the `json` object to parse the JSON content
         call json%initialize()

         ! Deserialize the JSON response(parsing the json)
         call json%deserialize(response%content)

         ! Traverse Repositories and Print Names

         ! Counter to traverse all repos one by one
         i = 1

         ! Storing the string equivalent of i in the count variable
         count = to_string(i)

         ! Fetching the name of the 1st GitHub repository, if it exists (found is set to true)
         call json%get('['//count//'].name', value, found)

         ! Enter the loop to traverse all repositories while they exist

>          do while(found)

Reviewer: This loop is strange. Does it call twice `json%get` for the same count? Why is the last `json%get` needed?

>          ! Traverse Repositories and Print Names
> 
>          ! Counter to traverse all repos one by one
>          i = 1
> 
>          ! Storing the string equivalent of i in the count variable
>          count = to_string(i)
> 
>          ! Fetching the name of the 1st GitHub repository, if it exists (found is set to true)
>          call json%get('['//count//'].name', value, found)
> 
>          ! Enter the loop to traverse all repositories while they exist
>          do while(found)
> 
>             ! Fetch the name of the current repository (based on the `i` counter) and check if it exists
>             call json%get('['//count//'].name', value, found)
> 
>             ! If the repository name exists (`found` is true), print the repository number and name
>             if (found) then
>                print *, count//'. ', value
>             end if
> 
>             ! Increment the counter for the next repository
>             i = i + 1
> 
>             ! Convert the updated counter to its string representation and store it in count variable
>             count = to_string(i)
> 
>             ! Fetch the name of the next repository (based on the updated `i` counter) and update `found` accordingly
>             call json%get('['//count//'].name', value, found)
> 
>          end do

Reviewer: Here is a suggestion (with a single call to `json%get`) that looks easier to understand (at least to me):

```suggestion
         ! Traverse Repositories and Print Names

         ! Counter to traverse all repos one by one
         i = 0

         ! Enter the loop to traverse all repositories while they exist
         do
            ! Increment the counter for the next repository
            i = i + 1

            ! Convert the updated counter to its string representation and store it in count variable
            count = to_string(i)

            ! Fetch the name of the current repository (based on the `i` counter) and check if it exists
            call json%get('['//count//'].name', value, found)

            if(.not.found) exit

            ! If the repository name exists (`found` is true), print the repository number and name
            print *, count//'. ', value            
            
         end do
```

Author: Thanks, @jvdp1 now it's more readable
